### PR TITLE
fix(jumplist): Ctrl+o, Ctrl+i behave unintuitively when deleting buffers

### LINKED ISSUE #25365: `Ctrl+o` and `Ctrl+i` behave unintuitively when deleting buffers
### Problem

I noticed that navigation with `Ctrl+i/o` is very weird after closing buffers that were part of the jump list.

To be honest, I'm not really sure if this is really a bug since `vim` behaves the same way and it seems to me like someone would have complained about this already if it was not intended behavior, but I couldn't find any issues or posts about this.

Regardless, this is something that really hinders my workflow (I use `oil.nvim` and every time I use it to navigate to a file it basically renders my jump list unusable) so I would really like to find a solution for my problem. 

### Steps to reproduce

I was able to reproduce this behavior with with `nvim --clean` on `0.10.0` and `stable`. I have added screenshots of the output of `:ju` in hopes of making it easier to follow.

**First case:**
- Open three buffers

- `Ctrl+o` to get to the second buffer

- `:bw` to close the buffer

     - Would expect `Ctrl+i` to move to the third buffer again, instead does nothing
- Press `Ctrl+o` to move to the first buffer

    - For some reason `Makefile` is now in front of `CHANGELOG.md`
    - `:ju` shows we are looking at `Makefile` but the `CHANGELOG.md` buffer is displayed in the window
- Press `Ctrl+i`, index is now on `-invalid-`, still the wrong buffer displayed

- Press `Ctrl+i` again, index is now at `CHANGELOG.md`

    - Jump list is now inverted from how it was in the beginning
    - From now on you have to press `Ctrl+o` or `Ctrl+i` twice whenever you want to move between these buffers because the first time the `-invalid-` buffer will be selected

**Second case:**
- Open three buffers

- `Ctrl+o` to get to the second buffer

- `:bw` to close the buffer

- Open a fourth buffer

    - Original jump list entries have switched as in the first case
- Press `Ctrl+o`, goes back to the (previously first) second buffer

- Press `Ctrl+o`, index is at `-invalid-` buffer

- Press `Ctrl+o` again, index is at `Makefile`

- Press `Ctrl+i`

    - Moves to `README`, completely skipping the `CHANGELOG.md`
    - From now on going from `README` to `Makefile` takes three presses of `Ctrl+o` (because of the invalid buffer) but going from `Makefile` to `README` takes only one press of `Ctrl+i`

I ran into multiple variations of this behavior during normal usage, some of them even making buffers that are in the jump list completely unreachable by using `Ctrl+i/o` , but I sadly I cannot reproduce them.

### Expected behavior

I would expect 
- navigation to always be symmetric (pressing `Ctrl+o` n times followed by `Ctrl+i` n times always gets me back to where I started and vice versa) 
- invalid buffers to either be skipped by `Ctrl+i/o` or removed completely from the jump list, to avoid having to press `Ctrl+i/o` multiple times to get to the next/previous buffer

### Neovim version (nvim -v)

0.10.0 commit 1635c9e75e21e07c4331cf983e14a11c7e09b119

### Vim (not Nvim) behaves the same?

yes

### Operating system/version

Arch Linux

### Terminal name/version

foot 1.15.3

### $TERM environment variable

xterm-256color

### Installation

build from repo

### PR DESCRIPTION
**Pull request by vE5li**
Closes #25365.

These changes fix the weird behavior described in #25365, by applying two simple changes:
- remove all entries of a buffer from the jump list when deleting it
- not adding a new entry to the jump list if the next buffer to be displayed was also found in the jump list

This makes the jump list behave exactly as you would expect when deleting buffers

### CODE REVIEW

File: src/nvim/buffer.c [hunk 1]
```diff
@@ -1178,6 +1178,32 @@ static int empty_curbuf(int close_others, int forceit, int action)
   return retval;
 }
 
+/// Remove every jump list entry referring to a given buffer.
+/// This function will also adjust the current jump list index.
+void remove_buffer_from_jumplist(buf_T *deleted_buffer)
```
Comment (justinmk) [inline src/nvim/buffer.c:1183]:
```suggestion
void remove_buffer_from_jumplist(buf_T *buf)
```

File: src/nvim/buffer.c [hunk 2]
```diff
@@ -1178,6 +1178,32 @@ static int empty_curbuf(int close_others, int forceit, int action)
...
+{
+  // Remove all jump list entries that match the deleted buffer.
+  for (int i = curwin->w_jumplistlen - 1; i >= 0; i--) {
+    buf_T *tested_buffer = buflist_findnr(curwin->w_jumplist[i].fmark.fnum);
```
Comment (justinmk) [inline src/nvim/buffer.c:1187]:
```suggestion
    buf_T *buf = buflist_findnr(curwin->w_jumplist[i].fmark.fnum);
```

**Comment by justinmk** [comment]
Thanks! Looks worth trying, but needs a test. Also see questions below.

Test could go in [test/functional/editor/jump_spec.lua](https://github.com/neovim/neovim/blob/master/test/functional/editor/jump_spec.lua). You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/editor/jump_spec.lua make functionaltest

## Questions

- would it make sense for the logic to be part of `cleanup_jumplist` ? or why not?
- is this related to https://github.com/neovim/neovim/pull/9805 , or why not?
    - note that was partially reverted in [21874](https://github.com/neovim/neovim/pull/21874), need to make sure we don't regress again

**Comment by vE5li** [comment]
I have some time the following days so I will try to add some tests.

To answer your questions:
- I don't think it makes sense, because of the way that `cleanup_jumplist` is used throughout the code base. The cleanup I'm doing here is exclusively applicable when deleting a buffer. However, I would like to continue cleaning up the jump list even after this PR is merged, and ideally, the `cleanup_jumplist` function can be removed completely, by making sure the jump list is never put into a "dirty" state to begin with. I think that doing so will make jump lists more stable, but more importantly, make the jump list logic much easier to maintain, as I noticed these `cleanup_jumplist` calls make it very hard to follow what state the jump list is in
- I don't really see #9805 as being related in any way other than it is also concerning the jump list. Given what I said above I would hope that both it and #21874 can be solved in a more fundamental way

**Comment by vE5li** [comment]
Tests are added :+1:

**Comment by vE5li** [comment]
@justinmk Would be great if you could take another look at this

**Comment by justinmk** [comment]
> I would like to continue cleaning up the jump list even after this PR is merged, and ideally, the `cleanup_jumplist` function can be removed completely, by making sure the jump list is never put into a "dirty" state to begin with. ... make the jump list logic much easier to maintain, as I noticed these `cleanup_jumplist` calls make it very hard to follow what state the jump list is in. ... and [revert: "shada/context: fully remove jumplist duplicates #10898" #21874](https://github.com/neovim/neovim/pull/21874) can be solved in a more fundamental way

💯  Very much looking forward to that.

File: test/functional/editor/jump_spec.lua [hunk 3]
```diff
@@ -182,6 +182,109 @@ describe("jumpoptions=stack behaves like 'tagstack'", function()
...
+  local content2 = base_content..'2'
+  local content3 = base_content..'3'
+  local function delete_buf(file)
+    local bufnr = funcs.bufnr(file)
+    command('bw '..bufnr)
+  end
+
+  before_each(function()
+    clear()
+    feed(':clearjumps<cr>')
```
Comment (justinmk) [inline test/functional/editor/jump_spec.lua:201]: is feed() intentional or can command be used? feed() is asynchronous and hides errors.

```suggestion
    command('clearjumps')
```

File: test/functional/editor/jump_spec.lua [hunk 4]
```diff
@@ -182,6 +182,109 @@ describe("jumpoptions=stack behaves like 'tagstack'", function()
...
+       .. '>  0     1    0 '..content3,
+       exec_capture('jumps'))
+  end)
+
+  it('deletes jump list entries when another buffer is deleted', function()
+    eq(   ''
+       .. ' jump line  col file/text\n'
+       .. '   2     1    0 '..file1..'\n'
+       .. '   1     1    0 '..file2..'\n'
+       .. '>',
```
Comment (justinmk) [inline test/functional/editor/jump_spec.lua:237]: could use dedent() with a literaly `[[...]]` string.

File: test/functional/editor/jump_spec.lua [hunk 5]
```diff
@@ -182,6 +182,109 @@ describe("jumpoptions=stack behaves like 'tagstack'", function()
...
+  local file1 = base_file..'1'
+  local file2 = base_file..'2'
+  local file3 = base_file..'3'
+  local base_content = 'text'
+  local content1 = base_content..'1'
+  local content2 = base_content..'2'
+  local content3 = base_content..'3'
+  local function delete_buf(file)
+    local bufnr = funcs.bufnr(file)
+    command('bw '..bufnr)
```
Comment (justinmk) [inline test/functional/editor/jump_spec.lua:196]: use full command names in code. bwipout also works with buffer names, is bufnr() needed?

```suggestion
    command('bwipeout '..file)
```

File: src/nvim/buffer.c [hunk 6]
```diff
@@ -1178,6 +1178,32 @@ static int empty_curbuf(int close_others, int forceit, int action)
   return retval;
 }
 
+/// Remove every jump list entry referring to a given buffer.
+/// This function will also adjust the current jump list index.
+void remove_buffer_from_jumplist(buf_T *deleted_buf)
```
Comment (justinmk) [inline src/nvim/buffer.c:1183]:
```suggestion
void buf_remove_from_jumplist(buf_T *deleted_buf)
```

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
> I would like to continue cleaning up the jump list even after this PR is merged, and ideally, the `cleanup_jumplist` function can be removed completely, by making sure the jump list is never put into a "dirty" state to begin with. ... make the jump list logic much easier to maintain, as I noticed these `cleanup_jumplist` calls make it very hard to follow what state the jump list is in. ... and [revert: "shada/context: fully remove jumplist duplicates #10898" #21874](https://github.com/neovim/neovim/pull/21874) can be solved in a more fundamental way

💯  Very much looking forward to that.

File: test/functional/editor/jump_spec.lua (hunk 5 above)
Comment (vE5li) [reply test/functional/editor/jump_spec.lua]: I didn't realize, thank you for the info!

Comment (vE5li) [reply test/functional/editor/jump_spec.lua]: I didn't realize, thanks for pointing it out!

File: test/functional/editor/jump_spec.lua (hunk 4 above)
Comment (vE5li) [reply test/functional/editor/jump_spec.lua:233]: How would I do that nicely given that I have to format the string inside the `[[...]]`? I thought it would be nice not to have the file name/content hard coded in the string 30 times to make the code easier to maintain, but if that style of creating the literals is not in line with the rest of the tests I can change it.

Comment (vE5li) [reply test/functional/editor/jump_spec.lua:233]: @justinmk I really don't want to be pushy and I know you're a busy guy but it would be really great if you could help me out here and tell me what else I need to do before merging

Comment (justinmk) [reply test/functional/editor/jump_spec.lua:233]: > I thought it would be nice not to have the file name/content hard coded in the string

definitely. sometimes it's worth using a trick like replacing a token, e.g. `(([[foo {file1} ...]]):gsub('%{file1%}', 'x'))`

don't need to block on this, though.

File: test/functional/editor/jump_spec.lua [hunk 7]
```diff
@@ -182,6 +182,105 @@ describe("jumpoptions=stack behaves like 'tagstack'", function()
...
+
+  it('sets the correct jump index when the current buffer is deleted', function()
+    feed('<C-O>')
+
+    eq(   ''
+       .. ' jump line  col file/text\n'
+       .. '   1     1    0 '..file1..'\n'
+       .. '>  0     1    0 '..content2..'\n'
+       .. '   1     1    0 '..file3,
+       exec_capture('jumps'))
```
Comment (justinmk) [inline test/functional/editor/jump_spec.lua:253]: there is also `:help getjumplist()`, would that be useful for checking this instead of a `:jumps` screentest?

Comment (justinmk) [inline test/functional/editor/jump_spec.lua:253]: there is also `:help getjumplist()`, would that be useful for checking this instead of a `:jumps` screentest?

(note: the existing tests in this file were added before `getjumplist()` was added).

```suggestion
    eq({ ... }, funcs.getjumplist())
```

Comment (vE5li) [reply test/functional/editor/jump_spec.lua:253]: I'm not sure to be honest. I like the fact that in the current state it is very easy to understand the state of the jump list, and to visually compare when running the same steps manually. On the other hand, this test now depends on the formatting that `jumps` uses, so if that is ever changed these tests would break too.

I would propose: we keep it like this for now and if it's a good idea I create a follow up PR where I migrate all the tests to use `getjumplist` instead, not just my own.

**Comment by vE5li** [comment]
> It looks like CI is failing, did you check those?

I did, but to be honest I am having trouble testing this project effectively. The tests fail for me locally and I'm struggling to understand the output of the tests/CI.

Take the output of `ubuntu-22.04 asan clang oldtest` for example:
```
Executed:  1313 Tests
 Skipped:   179 Tests
  Failed:     0 Tests

Failures: 
	Test test_quickfix failed, see output above and summary for more details
```
> Failed 0

but it still fails. And `ubuntu-22.04 asan clang functionaltest` logs
```
ERR 2023-12-10T12:50:29.388 T1346.12868.0 libuv_process_spawn:90: uv_spawn(/usr/bin/pwd) failed: permission denied
```
But I guess this test failing is just a consequence of the `oldtest` failing?

I will look into fixing the failing test(s) today :+1:

**Comment by vE5li** [comment]
Alright, I have fixed the issue and I squashed my commits to make the commit message linter happy.

From my point of view this is now ready to merge

**Comment by vE5li** [comment]
@justinmk I rebased on master again, please see if this can be merged

**Comment by vE5li** [comment]
@justinmk ping

**Comment by vE5li** [comment]
@zeertzjq Is it possible that you or someone else could take a look at this?

**Comment by justinmk** [comment]
@vE5li Thank you again, more intuitive jumplist will be great. 

I added a Problem/Solution description to the commit message based on your comments. That helps a lot for others to understand why a change was made, and for reviewers to understand if something is mergable.

> However, I would like to continue cleaning up the jump list even after this PR is merged, and ideally, the `cleanup_jumplist` function can be removed completely, by making sure the jump list is never put into a "dirty" state to begin with. I think that doing so will make jump lists more stable, but more importantly, make the jump list logic much easier to maintain, as I noticed these `cleanup_jumplist` calls make it very hard to follow what state the jump list is in

Eager to see the result of that :)

**Comment by Maltimore** [comment]
Hi all,
a very similar issue to this is that when deleting a buffer which used to be the alternate buffer, and then jumping to the alternate buffer (by default with with `CTRL-^` ), the deleted buffer gets loaded again.
Should I open a separate issue for that or is that the intended behavior?

**Comment by glepnir** [comment]
create an issue.

**Comment by Maltimore** [comment]
I opened https://github.com/neovim/neovim/issues/27285 to report the issue with the `alternate-file`.

**Comment by abilkhan024** [comment]
Hi! Is there a way to alter the behaviour of this change perhaps with a global vairable, because I was relying for this behaviour for a longest time, and as far as I know it is the default vim behavior. Or this change will be permanent default for neovim?

**Comment by justinmk** [comment]
> would like to continue cleaning up the jump list even after this PR is merged

@vE5li still interested in this?

**Comment by vE5li** [comment]
@justinmk Yes I am.

To be honest I was a bit discouraged to start working on it again by the process of merging my last PR. There was a period of around two months where I didn't get any response, even though I pinged multiple times.

I understand that maintaining a big open source repository such as Neovim is a lot of work, and I absolutely appreciate it, but not knowing what the state of your PR is can be very frustrating.

Especially if the directive is "Please keep your branch up to date with the current master". I had to resolve conflicts multiple times, just for my PR to be merged in the same state it was two month earlier.

I'm not trying to make a big deal out of it, I just hope that for the next PR we can find a way to communicate more quickly, so I can avoid some unnecessary work and frustration.

**Comment by justinmk** [comment]
@vE5li should go faster this time since there is no imminent release.
